src/data/blob.rs

```diff
@@ -0,0 +1,337 @@
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::{
+    fmt::{self, Debug, Formatter},
+    u64,
+};
+use tiny_keccak;
+
+/// Maximum allowed size for a serialised blob (ID) to grow to
+pub const MAX_BLOB_SIZE_IN_BYTES: u64 = 1024 * 1024 + 10 * 1024;
+
+/// PrivateBlob: an immutable chunk of data which can be deleted. Can only be fetched
+/// by the listed owner.
+#[derive(Hash, Eq, PartialEq, PartialOrd, Ord, Clone)]
+pub struct PrivateBlob {
+    /// Address.
+    address: Address,
+    /// Contained data.
+    value: Vec<u8>,
+    /// Contains a set of owners of this data. DataHandlers enforce that a DELETE or OWNED-GET type
+    /// of request is coming from the ClientHandler of the owners.
+    owner: PublicKey,
+}
+
+impl PrivateBlob {
+    /// Creates a new instance of `PrivateBlob`
+    pub fn new(value: Vec<u8>, owner: PublicKey) -> Self {
+        let hash_of_value = tiny_keccak::sha3_256(&value);
+        let serialised_contents = utils::serialise(&(hash_of_value, &owner));
+        let address = Address::Private(XorName(tiny_keccak::sha3_256(&serialised_contents)));
+
+        Self {
+            address,
+            value,
+            owner,
+        }
+    }
+
+    /// Returns the value.
+    pub fn value(&self) -> &Vec<u8> {
+        &self.value
+    }
+
+    /// Returns the set of owners.
+    pub fn is_owner(&self, user: PublicKey) -> bool {
+        self.owner == user
+    }
+
+    /// Returns the set of owners.
+    pub fn owner(&self) -> &PublicKey {
+        &self.owner
+    }
+
+    /// Returns the address.
+    pub fn address(&self) -> &Address {
+        &self.address
+    }
+
+    /// Returns the name.
+    pub fn name(&self) -> &XorName {
+        self.address.name()
+    }
+
+    /// Returns size of contained value.
+    pub fn payload_size(&self) -> usize {
+        self.value.len()
+    }
+
+    /// Returns size of this BlobData after serialisation.
+    pub fn serialised_size(&self) -> u64 {
+        serialized_size(self).unwrap_or(u64::MAX)
+    }
+
+    /// Return true if the size is valid
+    pub fn valid_size(&self) -> bool {
+        self.serialised_size() <= MAX_BLOB_SIZE_IN_BYTES
+    }
+}
+
+impl Serialize for PrivateBlob {
+    fn serialize<S: Serializer>(&self, serialiser: S) -> Result<S::Ok, S::Error> {
+        (&self.value, &self.owner).serialize(serialiser)
+    }
+}
+
+impl<'de> Deserialize<'de> for PrivateBlob {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        let (value, owner): (Vec<u8>, PublicKey) = Deserialize::deserialize(deserializer)?;
+        Ok(PrivateBlob::new(value, owner))
+    }
+}
+
+impl Debug for PrivateBlob {
+    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
+        // TODO: Output owners?
+        write!(formatter, "PrivateBlob {:?}", self.name())
+    }
+}
+
+/// An immutable chunk of data which cannot be deleted.
+#[derive(Hash, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct PublicBlob {
```

Comment:
Mm, perhaps, but it would break the pattern used for Map and Sequence(in code currently called MapData and SequenceData).

```
/// Object storing a Map variant.
pub enum Map {
    /// Public instance with concurrency control.
    PublicSentried(PublicSentriedMap),
    /// Public instance.
    Public(PublicMap),
    /// Private instance with concurrency control.
    PrivateSentried(PrivateSentriedMap),
    /// Private instance.
    Private(PrivateMap),
}
```

I'd prefer keeping the pattern between the data types for expressing the same kind of phenomenon.